### PR TITLE
Moving Kafka dependencies out into version specific modules

### PR DESCRIPTION
**Pull request by shirshanka**
@ibuenros @pcadabam : This is probably not the final PR, but would like to get a review to see if things look similar to what we discussed. 

Tests are passing.

### CODE REVIEW

**Comment by chavdar** [comment]
@shirshanka can you describe the change at a high level?

**Comment by shirshanka** [comment]
@chavdar Travis seems to be happy.

**Comment by shirshanka** [comment]
@chavdar : Care to review now? 
I have tried to extract as much code as possible into gobblin-kafka-common, leaving bare minimum in the 08 and 09 directories. 

I'm not sure why the coverage build is failing. 
I do want to add an integration test for 09 to make sure that the writer works with an 09 cluster.

**Comment by shirshanka** [comment]
High-level summary of changes: 

1) A top-level directory called gobblin-modules which will be used to store any datasystem specific integrations such as kafka. 
2) gobblin-modules/gobblin-kafka-common: a module that contains classes that are related to Kafka integration but do not depend on Kafka. 
3) gobblin-modules/gobblin-kafka-08: classes that allow reading, writing to Kafka 0.8 and reporting metrics to Kafka 0.8
4) gobblin-modules/gobblin-kafka-09: future home of readers, writers, reporters for Kafka 0.9. Currently only hosts the KafkaWriter.
5) gobblin-core-base: datasystem independent classes extracted from gobblin-core that help in creating datasystem integrations
6) gobblin-metrics : slimmed down and no longer depends on kafka or bundles the kafka reporter
7) gobblin-core: to preserve backwards compatibility, pulls in gobblin-kafka-08 using the standard flavored build for gobblin. This module can only depend on either gobblin-kafka-08 or gobblin-kafka-09. This can be controlled through the -DgobblinFlavor build argument and the respective gradle file (e.g. gobblin-core/gobblin-flavor-standard.gradle, gobblin-core/gobblin-flavor-full.gradle, ..)

**Comment by shirshanka** [comment]
@abti : What better time to check in massive changes than Fri midnight?

File: build.gradle [hunk 1]
```diff
@@ -47,6 +47,15 @@ idea.project {
   languageLevel = JavaVersion.VERSION_1_7
 }
 
+// Gobblin build customization
+if (!project.hasProperty('gobblinFlavor')) {
```
Comment (abti) [inline build.gradle]: It seems there are a few distinct supported flavors, can we document them somewhere to be explicit about what version does what?

File: gobblin-metrics/src/main/java/gobblin/metrics/GobblinMetrics.java [hunk 2]
```diff
@@ -728,33 +680,43 @@ private void buildCustomMetricReporters(Properties properties) {
...
+        com.codahale.metrics.ScheduledReporter scheduledReporter = this.codahaleReportersCloser
+            .register(customCodahaleReporterFactory.newScheduledReporter(RootMetricContext.get(), properties));
+        String reporterSinkMsg = reporterSink.isPresent()?"to " + reporterSink.get():"";
+        LOGGER.info("Will start reporting metrics " + reporterSinkMsg + " using " + reporterClass);
+        this.codahaleScheduledReporters.add(scheduledReporter);
+
+      } else if (CustomReporterFactory.class.isAssignableFrom(clazz)) {
+        CustomReporterFactory customReporterFactory = ((CustomReporterFactory) clazz.getConstructor().newInstance());
+        customReporterFactory.newScheduledReporter(properties);
+
```
Comment (abti) [inline gobblin-metrics/src/main/java/gobblin/metrics/GobblinMetrics.java]: nit: a log message for easier tracking that this custom reporter was initialized

File: build.gradle (hunk 1 above)
Comment (shirshanka) [inline build.gradle]: Probably makes sense to add this in build documentation and integration docs for kafka etc. 
Will open separate PR for that later.

**Pull request merged**